pages/common/views/components/datatable.jsx

```diff
@@ -1,48 +1,125 @@
+      deleteLink = 'Remove'
+    } = this.props;
+    if (isUndefined(data)) {
+      throw new Error('data must be provided');
+    }
+    const columns = !isEmpty(schema)
+      ? pickBy(schema, s => !s.expandable)
+      : Object.keys(data[0]).reduce((obj, key) => ({ ...obj, [key]: {} }), {});
+    const expandableData = pickBy(schema, s => s.expandable);
+    return (
+      <table className="govuk-react-datatable">
+        <thead>
+          <tr>
+            {
+              map(columns, (column, key) =>
+                <DatatableHeader key={key} id={key} sortable={sortable} { ...column } />
+              )
+            }
+          </tr>
+          <tr id="filter-header"></tr>
+        </thead>
+        <tbody>
           {
-            map(columns, (column, key) =>
-              <DatatableHeader key={key} id={key} sortable={sortable} { ...column } />
-            )
+            data.map(row => (
+              <Fragment key={row.id}>
+                <tr
+                  onClick={ expandable && (() => this.toggleContent(row.id)) }
+                  className={classnames({
+                    expandable,
+                    restrictions: row.notes,
+                    expanded: expandable && this.state && this.state.expanded[row.id]
+                  })}
+                >
+                  {
+                    map(columns, (schema, key) => {
+                      const datum = getValue({ row, schema, key });
+                      return <td key={key} className={key}>
+                        { schema.format ? schema.format(datum, row) : datum }
+                      </td>;
+                    })
+                  }
+                </tr>
+                {
+                  this.isExpandable(row.id) && (
+                    <tr className='expanded-content' onClick={ expandable && (() => this.toggleContent(row.id)) }>
+                      <td className="controls" colSpan={row.notes ? 1 : size(columns)}>
+                        <a href={`${url}/${row.id}/edit`}>{ editLink }</a>
+                        <a href={`${url}/${row.id}/delete`}>{ deleteLink }</a>
+                      </td>
+                      {
+                        Object.keys(expandableData).filter(r => row[r]).length > 0 && (
+                          <td colSpan={size(columns) - 1}>
+                            <dl>
+                              {
+                                Object.keys(expandableData).map(key => row[key] &&
+                                  <Fragment key={key}>
+                                    <dt>{schema[key].title || key}</dt>
+                                    <dd>{row[key]}</dd>
+                                  </Fragment>
+                                )
+                              }
+                            </dl>
+                          </td>
+                        )
+                      }
+                    </tr>
+                  )
+                }
+              </Fragment>
```

Comment:
Having all of this content hard-coded to fit this context feels overly specific. I'd be temped to just have a prop of `expandableContent` which attaches a component to a row, and that component gets rendered into the expanding row. Then we're not coupled to the "edit and delete links with a definition list" as the only thing we can put in an expansion.

Reply:
Maybe something in the formatters for an `expandable` property, that every row calls, and then if it returns anything, that gets rendered into the expansion, and if it returns null/false/undefined then it's not expandable?

i.e.
```js
const formatters = {
  ...
  expansion: row => {
    return <Fragment>
      ...
    </Fragment>
  }
}
```

Reply:
Obviously that's all fun and games until you have a column called `expansion` ;-)

Reply:
Hmm, the _expandable_ property isn't necessarily coupled with the schema though, as we show our edit and delete buttons there regardless of if there is any hidden content to show, the whole table gets an expandable prop, and the hidden/expandable field(s) are a separate concern

Reply:
unless we just add the edit/delete links inline and only have the rows with hidden content expandable (personal preference)

Reply:
whatever we do will have to be the same across all list pages, if we have an editable people list which doesn't have any hidden content would it still make sense to have expandable rows with just the edit/delete buttons inside?

Reply:
That was mostly just me covering the case where only some rows would have expandable content. In this particular case, the `Fragment` returned would always exist, and have content, it's just that some of the content would be conditional.

Reply:
Maybe munging it into the formatters isn't the best approach, and just an extra `renderExpansion` prop passed into the data table would be better.

Reply:
yeah agree - I'll pull that out. I'm just getting ahead of myself.

Reply:
> Maybe munging it into the formatters isn't the best approach, and just an extra renderExpansion prop passed into the data table would be better.

a la https://github.com/UKHomeOffice/asl-service/pull/38/files#diff-572f56089e3d09ed2e0b0921a918c3aeR44 ?